tao_compiler/mlir/disc/transforms/disc_comp_intens_fusion_to_cuda_source.cc

```diff
@@ -0,0 +1,631 @@
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mlir-hlo/Dialect/lhlo/IR/lhlo_ops.h"
+#include "mlir-hlo/Dialect/mhlo/IR/hlo_ops.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Pass/Pass.h"
+#include "tensorflow/compiler/mlir/disc/IR/lhlo_disc_ops.h"
+#include "tensorflow/compiler/mlir/disc/disc_util.h"
+#include "tensorflow/compiler/mlir/disc/transforms/PassDetail.h"
+#include "tensorflow/compiler/mlir/disc/utils/source_emitter.h"
+
+namespace mlir {
+namespace disc_ral {
+namespace {
+
+constexpr const char* kSpecializedClassName = "SpecializedGemmFusion";
+constexpr const char* kSpecializedEpilogue = "SpecializedEpilogue";
+constexpr const char* kEpilogueIsHeavy = "EpilogueIsHeavy";
+
+constexpr const char* kGRank = "GRank";
+constexpr const char* kElementAType = "ElementAType";
+constexpr const char* kElementALayout = "ElementALayout";
+constexpr const char* kElementBType = "ElementBType";
+constexpr const char* kElementBLayout = "ElementBLayout";
+constexpr const char* kElementOutputType = "ElementOutputType";
+constexpr const char* kElementOutputLayout = "ElementOutputLayout";
+constexpr const char* kElementAccumulatorType = "ElementAccumulatorType";
+constexpr const char* kOperatorClassType = "OperatorClassType";
+constexpr const char* kSMArch = "SMArch";
+
+constexpr const char* kScaleKind = "EpilogueScaleKind";
+constexpr const char* kCountVectorized = "EpilogueCountVectorized";
+constexpr const char* kEpilogueType = "EpilogueElementType";
+
+constexpr const char* kGatherA = "IsGatherA";
+constexpr const char* kGatherB = "IsGatherB";
+constexpr const char* kScatterD = "IsScatterD";
+constexpr const char* kPermuteDLayout = "EpiloguePermuteDLayout";
+
+constexpr const char* kParameterPermute = "ParameterPermute";
+
+constexpr const char* kGemmFusionFuncName = "gemmFusionFunc";
+
+int64_t stringReplaceInplace(std::string& subject, const std::string& oldsub,
+                             const std::string& newsub, bool replace_all) {
+  if (oldsub.empty()) {
+    return 0;
+  }
+  int64_t count = 0;
+  size_t pos = 0;
+  while ((pos = subject.find(oldsub, pos)) != std::string::npos) {
+    subject.replace(pos, oldsub.size(), newsub);
+    count++;
+    pos += newsub.size();
+    if (!replace_all) {
+      break;
+    }
+  }
+  return count;
+}
+
+struct DiscCompIntensFusionToCUDASourcePass
+    : public DiscCompIntensFusionToCUDASourcePassBase<
+          DiscCompIntensFusionToCUDASourcePass> {
+ public:
+  DiscCompIntensFusionToCUDASourcePass() = delete;
+  explicit DiscCompIntensFusionToCUDASourcePass(int cc_major, int cc_minor) {
```

Comment:
nit: seems no need to add `= delete` since the default one won't created if there's any user-defined constructors.
nit: add `explicit` to constructor with multiple argument makes no sense.